### PR TITLE
Feature/tags for elasticsearchwriter

### LINKED ISSUE #6837: Support host and service templates to send variables to Elasticsearch
It would be great if we could select host.vars and service.vars additional sent to Elasticsearch using the [ElastichsearchWriter.](https://github.com/Icinga/icinga2/blob/master/doc/09-object-types.md#elasticsearchwriter) The idea was inspired by #6213.

This should follow the template logic of [InfluxdbWriter](https://github.com/Icinga/icinga2/blob/master/doc/09-object-types.md#objecttype-influxdbwriter):
```
object ElasticsearchWriter "elasticsearch" {
  # other options...

  host_template = {
    tags = {
      location = "$host.vars.geodata$"
    }
  }
  service_template = {
    tags = {
      snmp_oid = "$service.vars.snmpv3_oid$"
    }
  }
}
```
See also [Tags vs Fields](https://discuss.elastic.co/t/tags-vs-fields/24014)

### PR DESCRIPTION
**Pull request by SebastianOpeni**
Offers to add additional tags to entries written by the ElasticSearchWriter.

As discussed with Eric at the icinga summit we are submitting this feature. Please revise the code and let us know if there is something we need to improve. otherwise we would be happy for timely merge.

fixes #6837

### CODE REVIEW

**Review by ngoeddel-openi — APPROVED** [review]
lgtm

**Comment by SebastianOpeni** [comment]
The CLA should be signed now.

**Comment by bobapple** [comment]
@cla-bot check

File: lib/perfdata/elasticsearchwriter.cpp [hunk 1]
```diff
@@ -131,6 +132,36 @@ void ElasticsearchWriter::Pause()
 	ObjectImpl<ElasticsearchWriter>::Pause();
 }
 
+void ElasticsearchWriter::AddTemplateTags(const Dictionary::Ptr& fields, const Host::Ptr& host, const Service::Ptr& service, const CheckResult::Ptr& cr)
+{
+	MacroProcessor::ResolverList resolvers;
```
Comment (yhabteab) [inline lib/perfdata/elasticsearchwriter.cpp:137]: You can declare this above or below the `ObjectLock olock ...` within the if block instead.

**Review by yhabteab — CHANGES_REQUESTED** [review]
Hi, thanks for your contribution! I've just been looking over the code changes and left some inline comments below, but haven't tested it with an Elastic/OpenSearch instance yet. Other than that, can you please rebase this against the current master to fix the GitHub actions. Thanks!

**Comment by SebastianOpeni** [comment]
Hi thanks for your review :)
I am currently working on your remarks but might have some questions beforehand.

File: lib/perfdata/elasticsearchwriter.hpp [hunk 2]
```diff
@@ -37,6 +40,7 @@ class ElasticsearchWriter final : public ObjectImpl<ElasticsearchWriter>
 	std::mutex m_DataBufferMutex;
 
 	void AddCheckResult(const Dictionary::Ptr& fields, const Checkable::Ptr& checkable, const CheckResult::Ptr& cr);
+	void AddTemplateTags(const Dictionary::Ptr& fields, const Host::Ptr& host, const Service::Ptr& service, const CheckResult::Ptr& cr);
```
[the remark this answers is not in the document]
Comment (SebastianOpeni) [reply lib/perfdata/elasticsearchwriter.hpp:43]: That is true and I also thought about doing so.
My reasoning was the following:
For the Resolver inside of the method I need to get the Host::Ptr and Service::Ptr from the Checkable::Ptr like it is done in other places.
But at all sources where AddTemplateTags will be called from this Host and Service retrieval has been done already just a couple of lines above.
So I would do the same "conversion" multiple times and therefore opted to call the Method with Host::Ptr and Service::Ptr directly.
If you would prefer the Checkable::Ptr way I can also implement it that way.

File: lib/perfdata/elasticsearchwriter.ti [hunk 3]
```diff
@@ -29,6 +29,18 @@ class ElasticsearchWriter : ConfigObject
...
+			return new Dictionary({
+			});
+		}}}
+	};
+	[config, required] Dictionary::Ptr service_tags_template {
+		default {{{
+			return new Dictionary({
+			});
+		}}}
+	};
```
[the remark this answers is not in the document]
Comment (SebastianOpeni) [reply lib/perfdata/elasticsearchwriter.ti:43]: You are absolutely right about the `required` keyword and I will remove it. :see_no_evil: 
(To be perfectly honest I took heavy inspiration for this part from the definition of the influx perfdata writer. So most likely it has to be removed there too.)

[the remark this answers is not in the document]
Comment (yhabteab) [reply lib/perfdata/elasticsearchwriter.ti:43]: > (To be perfectly honest I took heavy inspiration for this part from the definition of the influx perfdata writer. So most likely it has to be removed there too.)

The InfluxDB Writer doesn't return an empty dictionary by default, instead it requires that the two dictionary keys `measurement` and `tags` to be always there, i.e. the attribute `host_template` is never allowed to be empty. Here on the other hand, there are no really meaningful default values, so it should not be marked as required and the empty default value should be removed as well.

[the remark this answers is not in the document]
Comment (yhabteab) [reply lib/perfdata/elasticsearchwriter.ti:43]: If it is indeed supported, then you probably need to adjust validation errors, i.e. notice that in my example config snippet the opening `$` is missing and when trying to produce an error message, you use the `+` operator to concatenate `string` with the nested dictionary (`"Closing $ not found in macro format string '" + pair.second` and `pair.second` is the nested dictionary in this case).

File: lib/perfdata/elasticsearchwriter.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (yhabteab) [reply lib/perfdata/elasticsearchwriter.hpp:43]: > If you would prefer the Checkable::Ptr way I can also implement it that way.

The thing is, currently it looks like as the service parameter is always required and implies that it may never be a `nullptr`, thus I would either use `Checkable::ptr` instead and do the one-line casting again (my preferred variant) or make the `service` parameter optional.

File: .mailmap [hunk 4]
```diff
@@ -63,3 +63,4 @@ Robin O'Brien <robin@labs.epiuse.com> <robinjohnobrien@gmail.com>
 Roman Gerhardt <roman.gerhardt@cbc-x.com> <roman.gerhardt@cbc-x.com>
 Sebastian Chrostek <sebastian@chrostek.net> <sebastian@chrostek.net>
 Thomas Gelf <thomas.gelf@icinga.com> <thomas@gelf.net>
+Sebastian Grund <s.grund@openinfrastructure.de> <s.grund@openinfrastructure.de>
```
[the remark this answers is not in the document]
Comment (SebastianOpeni) [reply .mailmap]: I made sure to add a personal Mail-Address if I would stop being reachable under the work one.

File: doc/21-development.md
[the remark this answers is not in the document]
Comment (SebastianOpeni) [reply doc/21-development.md:1]: Would you prefer a new PullRequest for that?
Additionally I would also like to commit a minimal change: The influxdbwriter was also missing the `+ "'."` for its Validators.

File: lib/perfdata/elasticsearchwriter.hpp [hunk 5]
```diff
@@ -23,6 +23,9 @@ class ElasticsearchWriter final : public ObjectImpl<ElasticsearchWriter>
 
 	static String FormatTimestamp(double ts);
 
+	void ValidateHostTagsTemplate(const Lazy<Dictionary::Ptr> &lvalue, const ValidationUtils &utils);
+	void ValidateServiceTagsTemplate(const Lazy<Dictionary::Ptr> &lvalue, const ValidationUtils &utils);
```
[the remark this answers is not in the document]
Comment (SebastianOpeni) [reply lib/perfdata/elasticsearchwriter.hpp]: Done now

File: lib/perfdata/elasticsearchwriter.ti (hunk 3 above)
[the remark this answers is not in the document]
Comment (SebastianOpeni) [reply lib/perfdata/elasticsearchwriter.ti]: I removed both the default dictionary and the `required` mark

[the remark this answers is not in the document]
Comment (SebastianOpeni) [reply lib/perfdata/elasticsearchwriter.ti]: I decided that key-value pairs and one-level arrays of strings should be enough.
(More complex tagging can still be implemented via the usage of makros if needed.)

File: lib/perfdata/elasticsearchwriter.cpp [hunk 6]
```diff
@@ -683,3 +720,31 @@ String ElasticsearchWriter::FormatTimestamp(double ts)
...
+void ElasticsearchWriter::ValidateHostTagsTemplate(const Lazy<Dictionary::Ptr>& lvalue, const ValidationUtils& utils)
+{
+	ObjectImpl<ElasticsearchWriter>::ValidateHostTagsTemplate(lvalue, utils);
+
+	Dictionary::Ptr tags = lvalue();
+	if (tags) {
+		ObjectLock olock(tags);
+		for (const Dictionary::Pair& pair : tags) {
+			if (!MacroProcessor::ValidateMacroString(pair.second))
+				BOOST_THROW_EXCEPTION(ValidationError(this, { "host_tags_template", pair.first }, "Closing $ not found in macro format string '" + pair.second));
```
[the remark this answers is not in the document]
Comment (SebastianOpeni) [reply lib/perfdata/elasticsearchwriter.cpp]: Added the missing ending.

File: lib/perfdata/elasticsearchwriter.cpp [hunk 7]
```diff
@@ -131,6 +132,42 @@ void ElasticsearchWriter::Pause()
...
+			for (const Dictionary::Pair& pair : tmpl) {
+				String missing_macro;
+				Value value = MacroProcessor::ResolveMacros(pair.second, resolvers, cr, &missing_macro);
+
+				if (missing_macro.IsEmpty()) {
+					tags->Set(pair.first, value);
+				}
+			}
+		}
+		fields->Set("tags", tags);
```
Comment (lippserd) [inline lib/perfdata/elasticsearchwriter.cpp:167]: I'm not quite sure if `tags` is the right field for this. Isn't it common that `tags` is just a simple list instead of key-value pairs? I would set a different field or make it configureable.

**Comment by SebastianOpeni** [comment]
Thanks for your answers in the review.
I tried to implement everything as requested.

File: lib/perfdata/elasticsearchwriter.cpp (hunk 7 above)
Comment (SebastianOpeni) [reply lib/perfdata/elasticsearchwriter.cpp:167]: `tags` was used because the influxdb-writer is using it.
The functionality stays the same so if you have a preference for a name I would change it accordingly.
(`labels` would come to mind as a alternative.)

File: .mailmap (hunk 4 above)
[the remark this answers is not in the document]
Comment (yhabteab) [reply .mailmap]: The `.mailmap` is used, for example, to map all your commits in this repo that you have previously committed with the mail `s.grund@openinfrastructure.de` to the new mail `development@zottel92.de` when, for example, you change your employer and push new commits with the new mail address here. For this reason, the mails are mapped in the following order: `Author Name <NEW MAIL> <OLD MAIL>`, but it is not necessary to do this in advance.

File: doc/21-development.md
[the remark this answers is not in the document]
Comment (yhabteab) [reply doc/21-development.md:1]: > Would you prefer a new PullRequest for that? Additionally I would also like to commit a minimal change: The influxdbwriter was also missing the `+ "'."` for its Validators.

That would be great 👍.

**Review by yhabteab — CHANGES_REQUESTED** [review]

File: .mailmap (hunk 4 above)
[the remark this answers is not in the document]
Comment (SebastianOpeni) [reply .mailmap]: I see. I removed my entry :+1:

File: doc/21-development.md
[the remark this answers is not in the document]
Comment (SebastianOpeni) [reply doc/21-development.md:1]: I created the "closing quote" PullRequest here: https://github.com/Icinga/icinga2/pull/10174
The development doc problem was already found by another issue and got fixed in the meantime :)

File: lib/perfdata/elasticsearchwriter.cpp (hunk 7 above)
Comment (SebastianOpeni) [reply lib/perfdata/elasticsearchwriter.cpp:165]: I gave it some thought and I could also simply allow tags to be of type array.
That would change some of the logic we had reviewed until now but if that would be preferred by you, I could also implement it that way.

Comment (Al2Klimov) [reply lib/perfdata/elasticsearchwriter.cpp:165]: > Isn't it common that `tags` is just a simple list instead of key-value pairs?

I'm afraid, in perfdata DB context, they indeed are key-value pairs:

* https://docs.influxdata.com/influxdb/v1/concepts/glossary/#tag-key
* https://graphite.readthedocs.io/en/latest/tags.html

**Review by Al2Klimov — DISMISSED** [review]
## Tests

TL;DR: It works! 👍

```
➜  ~ curl -fsS http://127.0.0.1:9200/icinga2-2024.09.25/_search/ |python3 -mjson.tool
{
    "took": 5,
    "timed_out": false,
    "_shards": {
        "total": 1,
        "successful": 1,
        "skipped": 0,
        "failed": 0
    },
    "hits": {
        "total": {
            "value": 6,
            "relation": "eq"
        },
        "max_score": 1.0,
        "hits": [
            {
                "_index": "icinga2-2024.09.25",
                "_type": "_doc",
                "_id": "pAd1KZIB9Wwi17GmLApi",
                "_score": 1.0,
                "_source": {
                    "@timestamp": "2024-09-25T15:53:07.855+0200",
                    "check_command": "hostalive",
                    "check_result.check_source": "ws-aklimov7777777.local",
                    "check_result.command": [
                        "/Users/aklimov/NET/WS/icinga2/prefix/usr/lib/nagios/plugins/check_ping",
                        "-H",
                        "127.0.0.1",
                        "-c",
                        "5000,100%",
                        "-w",
                        "3000,80%"
                    ],
                    "check_result.execution_end": "2024-09-25T15:53:07.855+0200",
                    "check_result.execution_start": "2024-09-25T15:53:07.854+0200",
                    "check_result.execution_time": 0.0009140968322753906,
                    "check_result.exit_status": 128,
                    "check_result.latency": 0.0381932258605957,
                    "check_result.output": "execvpe(/Users/aklimov/NET/WS/icinga2/prefix/usr/lib/nagios/plugins/check_ping) failed: No such file or directory<Terminated with exit code 128 (0x80).>",
                    "check_result.schedule_end": "2024-09-25T15:53:07.855+0200",
                    "check_result.schedule_start": "2024-09-25T15:53:07.816+0200",
                    "check_result.state": 3,
                    "check_result.vars_after": {
                        "attempt": 2,
                        "reachable": true,
                        "state": 3,
                        "state_type": 0
                    },
                    "check_result.vars_before": null,
                    "current_check_attempt": 2,
                    "host": "ws-aklimov7777777.local",
                    "last_hard_state": 1,
                    "last_state": 1,
                    "max_check_attempts": 3,
                    "reachable": true,
                    "state": 1,
                    "state_type": 0,
                    "tags": {
                        "a_list_of_items": [
                            "foo",
                            "bar"
                        ],
                        "custom_label": "A Custom Label",
                        "os_name": "Linux"
                    },
                    "timestamp": "2024-09-25T15:53:07.855+0200",
                    "type": "icinga2.event.checkresult"
                }
            },
            {
                "_index": "icinga2-2024.09.25",
                "_type": "_doc",
                "_id": "pQd1KZIB9Wwi17GmLApj",
                "_score": 1.0,
                "_source": {
                    "@timestamp": "2024-09-25T15:53:07.855+0200",
                    "check_command": "hostalive",
                    "check_result.check_source": "ws-aklimov7777777.local",
                    "check_result.command": [
                        "/Users/aklimov/NET/WS/icinga2/prefix/usr/lib/nagios/plugins/check_ping",
                        "-H",
                        "127.0.0.1",
                        "-c",
                        "5000,100%",
                        "-w",
                        "3000,80%"
                    ],
                    "check_result.execution_end": "2024-09-25T15:53:07.855+0200",
                    "check_result.execution_start": "2024-09-25T15:53:07.854+0200",
                    "check_result.execution_time": 0.0009140968322753906,
                    "check_result.exit_status": 128,
                    "check_result.latency": 0.0381932258605957,
                    "check_result.output": "execvpe(/Users/aklimov/NET/WS/icinga2/prefix/usr/lib/nagios/plugins/check_ping) failed: No such file or directory<Terminated with exit code 128 (0x80).>",
                    "check_result.schedule_end": "2024-09-25T15:53:07.855+0200",
                    "check_result.schedule_start": "2024-09-25T15:53:07.816+0200",
                    "check_result.state": 3,
                    "check_result.vars_after": {
                        "attempt": 2,
                        "reachable": true,
                        "state": 3,
                        "state_type": 0
                    },
                    "check_result.vars_before": null,
                    "current_check_attempt": 2,
                    "host": "ws-aklimov7777777.local",
                    "last_hard_state": 1,
                    "last_state": 1,
                    "max_check_attempts": 3,
                    "state": 1,
                    "tags": {
                        "a_list_of_items": [
                            "foo",
                            "bar"
                        ],
                        "custom_label": "A Custom Label",
                        "os_name": "Linux"
                    },
                    "timestamp": "2024-09-25T15:53:07.855+0200",
                    "type": "icinga2.event.statechange"
                }
            },
            {
                "_index": "icinga2-2024.09.25",
                "_type": "_doc",
                "_id": "pgd1KZIB9Wwi17GmLApj",
                "_score": 1.0,
                "_source": {
                    "@timestamp": "2024-09-25T15:53:10.689+0200",
                    "check_command": "disk",
                    "check_result.check_source": "ws-aklimov7777777.local",
                    "check_result.command": [
                        "/Users/aklimov/NET/WS/icinga2/prefix/usr/lib/nagios/plugins/check_disk",
                        "-c",
                        "10%",
                        "-w",
                        "20%",
                        "-X",
                        "none",
                        "-X",
                        "tmpfs",
                        "-X",
                        "sysfs",
                        "-X",
                        "proc",
                        "-X",
                        "configfs",
                        "-X",
                        "devtmpfs",
                        "-X",
                        "devfs",
                        "-X",
                        "mtmfs",
                        "-X",
                        "tracefs",
                        "-X",
                        "cgroup",
                        "-X",
                        "fuse.*",
                        "-X",
                        "fuse.gvfsd-fuse",
                        "-X",
                        "fuse.gvfs-fuse-daemon",
                        "-X",
                        "fuse.portal",
                        "-X",
                        "fuse.sshfs",
                        "-X",
                        "fdescfs",
                        "-X",
                        "overlay",
                        "-X",
                        "nsfs",
                        "-X",
                        "squashfs",
                        "-p",
                        "/"
                    ],
                    "check_result.execution_end": "2024-09-25T15:53:10.689+0200",
                    "check_result.execution_start": "2024-09-25T15:53:10.684+0200",
                    "check_result.execution_time": 0.004788875579833984,
                    "check_result.exit_status": 128,
                    "check_result.latency": 0.015203237533569336,
                    "check_result.output": "execvpe(/Users/aklimov/NET/WS/icinga2/prefix/usr/lib/nagios/plugins/check_disk) failed: No such file or directory<Terminated with exit code 128 (0x80).>",
                    "check_result.schedule_end": "2024-09-25T15:53:10.690+0200",
                    "check_result.schedule_start": "2024-09-25T15:53:10.670+0200",
                    "check_result.state": 3,
                    "check_result.vars_after": {
                        "attempt": 2,
                        "reachable": true,
                        "state": 3,
                        "state_type": 0
                    },
                    "check_result.vars_before": null,
                    "current_check_attempt": 2,
                    "host": "ws-aklimov7777777.local",
                    "last_hard_state": 3,
                    "last_state": 3,
                    "max_check_attempts": 5,
                    "reachable": true,
                    "service": "disk /",
                    "state": 3,
                    "state_type": 0,
                    "tags": {
                        "a_list_of_items": [
                            "foo",
                            "bar"
                        ],
                        "custom_label": "A Custom Label",
                        "os_name": "Linux",
                        "sv_name": "disk /"
                    },
                    "timestamp": "2024-09-25T15:53:10.689+0200",
                    "type": "icinga2.event.checkresult"
                }
            },
            {
                "_index": "icinga2-2024.09.25",
                "_type": "_doc",
                "_id": "pwd1KZIB9Wwi17GmLApj",
                "_score": 1.0,
                "_source": {
                    "@timestamp": "2024-09-25T15:53:10.689+0200",
                    "check_command": "disk",
                    "check_result.check_source": "ws-aklimov7777777.local",
                    "check_result.command": [
                        "/Users/aklimov/NET/WS/icinga2/prefix/usr/lib/nagios/plugins/check_disk",
                        "-c",
                        "10%",
                        "-w",
                        "20%",
                        "-X",
                        "none",
                        "-X",
                        "tmpfs",
                        "-X",
                        "sysfs",
                        "-X",
                        "proc",
                        "-X",
                        "configfs",
                        "-X",
                        "devtmpfs",
                        "-X",
                        "devfs",
                        "-X",
                        "mtmfs",
                        "-X",
                        "tracefs",
                        "-X",
                        "cgroup",
                        "-X",
                        "fuse.*",
                        "-X",
                        "fuse.gvfsd-fuse",
                        "-X",
                        "fuse.gvfs-fuse-daemon",
                        "-X",
                        "fuse.portal",
                        "-X",
                        "fuse.sshfs",
                        "-X",
                        "fdescfs",
                        "-X",
                        "overlay",
                        "-X",
                        "nsfs",
                        "-X",
                        "squashfs",
                        "-p",
                        "/"
                    ],
                    "check_result.execution_end": "2024-09-25T15:53:10.689+0200",
                    "check_result.execution_start": "2024-09-25T15:53:10.684+0200",
                    "check_result.execution_time": 0.004788875579833984,
                    "check_result.exit_status": 128,
                    "check_result.latency": 0.015203237533569336,
                    "check_result.output": "execvpe(/Users/aklimov/NET/WS/icinga2/prefix/usr/lib/nagios/plugins/check_disk) failed: No such file or directory<Terminated with exit code 128 (0x80).>",
                    "check_result.schedule_end": "2024-09-25T15:53:10.690+0200",
                    "check_result.schedule_start": "2024-09-25T15:53:10.670+0200",
                    "check_result.state": 3,
                    "check_result.vars_after": {
                        "attempt": 2,
                        "reachable": true,
                        "state": 3,
                        "state_type": 0
                    },
                    "check_result.vars_before": null,
                    "current_check_attempt": 2,
                    "host": "ws-aklimov7777777.local",
                    "last_hard_state": 3,
                    "last_state": 3,
                    "max_check_attempts": 5,
                    "service": "disk /",
                    "state": 3,
                    "tags": {
                        "a_list_of_items": [
                            "foo",
                            "bar"
                        ],
                        "custom_label": "A Custom Label",
                        "os_name": "Linux",
                        "sv_name": "disk /"
                    },
                    "timestamp": "2024-09-25T15:53:10.689+0200",
                    "type": "icinga2.event.statechange"
                }
            },
            {
                "_index": "icinga2-2024.09.25",
                "_type": "_doc",
                "_id": "qAd1KZIB9Wwi17GmLApj",
                "_score": 1.0,
                "_source": {
                    "@timestamp": "2024-09-25T15:53:15.753+0200",
                    "check_command": "ping4",
                    "check_result.check_source": "ws-aklimov7777777.local",
                    "check_result.command": [
                        "/Users/aklimov/NET/WS/icinga2/prefix/usr/lib/nagios/plugins/check_ping",
                        "-4",
                        "-H",
                        "127.0.0.1",
                        "-c",
                        "200,15%",
                        "-w",
                        "100,5%"
                    ],
                    "check_result.execution_end": "2024-09-25T15:53:15.753+0200",
                    "check_result.execution_start": "2024-09-25T15:53:15.750+0200",
                    "check_result.execution_time": 0.002724885940551758,
                    "check_result.exit_status": 128,
                    "check_result.latency": 0.012976884841918945,
                    "check_result.output": "execvpe(/Users/aklimov/NET/WS/icinga2/prefix/usr/lib/nagios/plugins/check_ping) failed: No such file or directory<Terminated with exit code 128 (0x80).>",
                    "check_result.schedule_end": "2024-09-25T15:53:15.755+0200",
                    "check_result.schedule_start": "2024-09-25T15:53:15.739+0200",
                    "check_result.state": 3,
                    "check_result.vars_after": {
                        "attempt": 2,
                        "reachable": true,
                        "state": 3,
                        "state_type": 0
                    },
                    "check_result.vars_before": null,
                    "current_check_attempt": 2,
                    "host": "ws-aklimov7777777.local",
                    "last_hard_state": 3,
                    "last_state": 3,
                    "max_check_attempts": 5,
                    "reachable": true,
                    "service": "ping4",
                    "state": 3,
                    "state_type": 0,
                    "tags": {
                        "a_list_of_items": [
                            "foo",
                            "bar"
                        ],
                        "custom_label": "A Custom Label",
                        "os_name": "Linux",
                        "sv_name": "ping4"
                    },
                    "timestamp": "2024-09-25T15:53:15.753+0200",
                    "type": "icinga2.event.checkresult"
                }
            },
            {
                "_index": "icinga2-2024.09.25",
                "_type": "_doc",
                "_id": "qQd1KZIB9Wwi17GmLApj",
                "_score": 1.0,
                "_source": {
                    "@timestamp": "2024-09-25T15:53:15.753+0200",
                    "check_command": "ping4",
                    "check_result.check_source": "ws-aklimov7777777.local",
                    "check_result.command": [
                        "/Users/aklimov/NET/WS/icinga2/prefix/usr/lib/nagios/plugins/check_ping",
                        "-4",
                        "-H",
                        "127.0.0.1",
                        "-c",
                        "200,15%",
                        "-w",
                        "100,5%"
                    ],
                    "check_result.execution_end": "2024-09-25T15:53:15.753+0200",
                    "check_result.execution_start": "2024-09-25T15:53:15.750+0200",
                    "check_result.execution_time": 0.002724885940551758,
                    "check_result.exit_status": 128,
                    "check_result.latency": 0.012976884841918945,
                    "check_result.output": "execvpe(/Users/aklimov/NET/WS/icinga2/prefix/usr/lib/nagios/plugins/check_ping) failed: No such file or directory<Terminated with exit code 128 (0x80).>",
                    "check_result.schedule_end": "2024-09-25T15:53:15.755+0200",
                    "check_result.schedule_start": "2024-09-25T15:53:15.739+0200",
                    "check_result.state": 3,
                    "check_result.vars_after": {
                        "attempt": 2,
                        "reachable": true,
                        "state": 3,
                        "state_type": 0
                    },
                    "check_result.vars_before": null,
                    "current_check_attempt": 2,
                    "host": "ws-aklimov7777777.local",
                    "last_hard_state": 3,
                    "last_state": 3,
                    "max_check_attempts": 5,
                    "service": "ping4",
                    "state": 3,
                    "tags": {
                        "a_list_of_items": [
                            "foo",
                            "bar"
                        ],
                        "custom_label": "A Custom Label",
                        "os_name": "Linux",
                        "sv_name": "ping4"
                    },
                    "timestamp": "2024-09-25T15:53:15.753+0200",
                    "type": "icinga2.event.statechange"
                }
            }
        ]
    }
}
➜  ~
```

File: lib/perfdata/elasticsearchwriter.cpp (hunk 7 above)
Comment (lippserd) [reply lib/perfdata/elasticsearchwriter.cpp:165]: > I'm afraid, in perfdata DB context, they indeed are key-value pairs:

Please show me where we set a field called `tags` in one of them. Supporting tags in both just means adding key-value pairs, with no reference to a `tags` field, right?. But here we set the field where I wonder if users of Elasticsearch or similar would expect a list rather than key-value pairs. `tag_set` sounds more reasonable to me. Anyway, before we just merge this, I'd like to get an informed opinion.

Comment (SebastianOpeni) [reply lib/perfdata/elasticsearchwriter.cpp:165]: For Icingas influxdbcommonwriter this is done here:
https://github.com/Icinga/icinga2/blob/master/lib/perfdata/influxdbcommonwriter.cpp#L253

The graphite writer does only allow to configure templates for the host and service name.

Comment (lippserd) [reply lib/perfdata/elasticsearchwriter.cpp:165]: > For Icingas influxdbcommonwriter this is done here:
> https://github.com/Icinga/icinga2/blob/master/lib/perfdata/influxdbcommonwriter.cpp#L253

But `tags` never reaches Influx DB.

> The graphite writer does only allow to configure templates for the host and service name.

Otherwise there would not be a `tags` field either.

Comment (SebastianOpeni) [reply lib/perfdata/elasticsearchwriter.cpp:165]: Found it as well just after Commenting. :see_no_evil: 
That was my bad. 
I'm very sorry for the confusion.

Comment (SebastianOpeni) [reply lib/perfdata/elasticsearchwriter.cpp:165]: To prevent confusion about the Type of `tags`, would you prefer to:
- Rename the field to `tag_set` or
- Add the configured key-values directly into the elasticSearch-Entries as done in the influxdb case?
?

Comment (Al2Klimov) [reply lib/perfdata/elasticsearchwriter.cpp:165]: > To prevent confusion about the Type of `tags`, would you prefer to:
> 
> * Rename the field to `tag_set` or

I think a tag_set property would be rather confusing in ES/Kibana.

> * Add the configured key-values directly into the elasticSearch-Entries as done in the influxdb case?

This, in contrast, has the advantages:

* No `tags.` boilerplate in our ES data
* Consistency with InfluxDB

Comment (SebastianOpeni) [reply lib/perfdata/elasticsearchwriter.cpp]: After my changes the configured tags are now added to the root element.
That also allows for `tags = ["foo", "bar"]` if desired.

Notice:
Since the templated_tags are added after the default fields and the performance Values this change will also allow to overwrite them.

**Review by yhabteab — CHANGES_REQUESTED** [review]
I tested this today with ElasticSearch 7.17.25 and it seems to be working fine 👍🏼. Though , whilst going through the changes again, I just noticed two small things, see inline comments below.

Edit: While you're at it, please rebase this against the master branch to get the GHAs pass.

File: doc/14-features.md [hunk 8]
```diff
@@ -398,6 +396,24 @@ check_result.perfdata.<perfdata-label>.warn
...
+  custom_label = "A Custom Label"
+  a_list_of_items = [ "foo", "bar" ]
+}
+```
+Will in addition to the above mentioned lines also contain:
+```
+os_name = "Linux"
+custom_label = "A Custom Label"
+a_list_of_items = [ "foo", "bar" ]
+```
```
[the remark this answers is not in the document]
Comment (SebastianOpeni) [reply doc/14-features.md]: I did commit your suggestions.

File: lib/perfdata/elasticsearchwriter.cpp [hunk 9]
```diff
@@ -131,6 +132,40 @@ void ElasticsearchWriter::Pause()
...
+	tie(host, service) = GetHostService(checkable);
+
+	Dictionary::Ptr tmpl;
+	if (service) {
+		if (GetServiceTagsTemplate()) {
+			tmpl = static_pointer_cast<Dictionary>(GetServiceTagsTemplate()->ShallowClone());
+		}
+	} else if (GetHostTagsTemplate()) {
+		tmpl = static_pointer_cast<Dictionary>(GetHostTagsTemplate()->ShallowClone());
+	}
```
[the remark this answers is not in the document]
Comment (SebastianOpeni) [reply lib/perfdata/elasticsearchwriter.cpp]: After working on different stuff for some time I also had to reread into it and you are absolutely right here.

As stated earlier I worked with the influxdb writer as a model.
There the returned Template is also used to store the resolved values and so cloning makes sense.
But since we only ever read from the retrieved template we do not have to clone it.

**Comment by SebastianOpeni** [comment]
I think I worked on your requested changes, and you also marked them as resolved, but somehow they are still blocking the pipeline :thinking:

**Comment by yhabteab** [comment]
> I think I worked on all your requested changes, and you also marked them as resolved, but somehow they are still blocking the pipeline 🤔

Hi, yes, thanks for that! Unfortunately, the GitHub actions were changed again with https://github.com/Icinga/icinga2/pull/10253 just after you rebased this against the master branch. You just need to rebase and force push again :)!

**Comment by SebastianOpeni** [comment]
Thanks for your fast reply :)
I just pushed the rebased version.
I hope the tests run smoothly :)

**Comment by SebastianOpeni** [comment]
The checks seem to be not the problem, but rather github does not register that we closed all your previous existing change requests. I think I have to "re-request" a review from you. :thinking:

**Comment by yhabteab** [comment]
> The checks seem to be not the problem, but rather github does not register that we closed all your previously existing change requests. I think I have to "re-request" a review from you. 🤔

Hi, sorry for the delay! Unfortunately, there was _again_ some changes made to the GitHub actions :-), and I wanted to rebase and clean up the commit histories for you but I'm not allowed to push it into this PR :(.
Bildschirmfoto 2025-01-24 um 08 54 22

**Comment by SebastianOpeni** [comment]
Happy new year and nice to hear from you :smiley: 
Thank you for the effort, I invited you with write permissions I hope that will give you the necessary possibilities :)

**Comment by yhabteab** [comment]
> I invited you with write permissions I hope that will give you the necessary possibilities :)

Thanks! But it would have been enough to just check the [`Allow edits by maintainers`](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork) box.

**Review by yhabteab — APPROVED** [review]
LFTM!

**Comment by julianbrost** [comment]
@yhabteab Can you please state what you expect @lippserd to review here?

**Comment by yhabteab** [comment]
> @yhabteab Can you please state what you expect @lippserd to review here?

Since @lippserd requested a change in https://github.com/Icinga/icinga2/pull/10074#discussion_r1771206872 not to send the tags fields to elastic search, so I wanted to know whether the changed (flattened) version of those matches his expectation.

**Review by lippserd — APPROVED** [review]
